### PR TITLE
First-class project handling in MergeCartoix

### PR DESCRIPTION
**Pull request by wemrysi**
Adds projection extraction and reification during `MergeCartoix`, no semantic changes to plans.

[ch8302]

### CODE REVIEW

File: foundation/src/main/scala/quasar/contrib/scalaz/nel.scala [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+package quasar.contrib.scalaz
+
+import scala.annotation.unchecked.uncheckedVariance
+
+import scalaz.NonEmptyList
+import scalaz.Liskov.{<~<, refl}
+
+object nel {
+  implicit final class NonEmptyListOps[A](xs: NonEmptyList[A]) {
+    def widen[B](implicit ev: A <~< B): NonEmptyList[B] =
```
Comment (djspiewak) [inline foundation/src/main/scala/quasar/contrib/scalaz/nel.scala]: Every time I've gotten into this kind of state, I've just given up and leveraged the fact that `cats.data.NonEmptyList` is properly covariant. :-)

File: qsu/src/main/scala/quasar/qsu/minimizers/CStage.scala [hunk 2]
```diff
@@ -30,14 +30,15 @@ import matryoshka.data.free._
...
+import scalaz.{Equal, Show}
+import scalaz.Scalaz._
 
-private[minimizers] sealed trait CStage[T[_[_]]] extends Product with Serializable
+/**
+ * @tparam P path
+ * @tparam S struct
+ * @tparam O shift output
+ */
+private[minimizers] sealed trait CStage[T[_[_]], +P, +S, +O] extends Product with Serializable
```
Comment (djspiewak) [inline qsu/src/main/scala/quasar/qsu/minimizers/CStage.scala]: Rip straightforward types…

File: qsu/src/main/scala/quasar/qsu/minimizers/CStage.scala [hunk 3]
```diff
@@ -46,35 +47,34 @@ private[minimizers] object CStage {
    * Represents a divergence in the graph, as if millions of rows
    * cried out in terror and were suddenly silenced.
    */
-  final case class Join[T[_[_]]](
-      cartoix: SMap[Symbol, Cartouche[T]],
-      joiner: FreeMapA[T, CartoucheRef])
-      extends CStage[T]
+  final case class Join[T[_[_]], P, S, O](
+      cartoix: SMap[Symbol, Cartouche[T, P, S, O]],
+      joiner: FreeMapA[T, Symbol])
```
Comment (djspiewak) [inline qsu/src/main/scala/quasar/qsu/minimizers/CStage.scala]: Oh I think I missed this. Do we no longer need to reference `IncludeIds` nodes further up in a cartouche?

File: qsu/src/main/scala/quasar/qsu/minimizers/MergeCartoix.scala [hunk 4]
```diff
@@ -20,53 +20,61 @@ package minimizers
...
 
-  private implicit val MM: Monoid[SMap[Symbol, Cartouche]] =
-    MonoidK[SMap[Symbol, ?]].algebra[Cartouche]
+  private implicit val MM: Monoid[SMap[Symbol, Cartouche0]] =
+    MonoidK[SMap[Symbol, ?]].algebra[Cartouche0]
 
   private val SourceKey = "source"
   private val func = construction.Func[T]
 
+  private def absurd[A]: A = scala.sys.error("absurd!")
```
Comment (djspiewak) [inline qsu/src/main/scala/quasar/qsu/minimizers/MergeCartoix.scala]: FYI: `quasar.contrib.std.errorImpossible`

Not sure if we want to standardize on that.

File: qsu/src/main/scala/quasar/qsu/minimizers/MergeCartoix.scala [hunk 5]
```diff
@@ -20,53 +20,61 @@ package minimizers
...
   import QSUGraph.Extractors._
   import RecFreeS.RecOps
+  import MapFuncsCore.{IntLit, ProjectIndex, ProjectKey, StrLit}
 
-  type Cartouche = quasar.qsu.minimizers.Cartouche[T]
-  type CStage = quasar.qsu.minimizers.CStage[T]
+  type Cartouche0 = quasar.qsu.minimizers.Cartouche[T, Index, Hole, Retain]
+  type Cartouche1 = quasar.qsu.minimizers.Cartouche[T, Nothing, FreeMap, Output]
+  type CStage0 = quasar.qsu.minimizers.CStage[T, Index, Hole, Retain]
+  type CStage1 = quasar.qsu.minimizers.CStage[T, Nothing, FreeMap, Output]
```
Comment (djspiewak) [inline qsu/src/main/scala/quasar/qsu/minimizers/MergeCartoix.scala]: Oooooooh! Very creative

File: qsu/src/main/scala/quasar/qsu/minimizers/MergeCartoix.scala [hunk 6]
```diff
@@ -323,92 +452,98 @@ sealed abstract class MergeCartoix[T[_[_]]: BirecursiveT: EqualT: RenderTreeT: S
...
 
-                case ((c, r), kv) => (c + kv, r)
+                // Cartouche doesn't terminate here, retain
+                //
+                // TODO: This isn't quite correct, see
+                //       https://app.clubhouse.io/data/story/9672/compatible-operations-to-id-and-value-components-of-a-shift-are-incorrectly-coalesced
+                //
+                //       Ideally when the cartouche did reference ids and didn't terminate,
+                //       we'd indicate that it should project the upstream id, not the value,
+                //       but we don't have a way to express this currently.
```
Comment (djspiewak) [inline qsu/src/main/scala/quasar/qsu/minimizers/MergeCartoix.scala]: Wasn't this exact expression what `CartoucheRef` *used* to achieve?

File: qsu/src/main/scala/quasar/qsu/minimizers/MergeCartoix.scala [hunk 7]
```diff
@@ -514,105 +709,109 @@ sealed abstract class MergeCartoix[T[_[_]]: BirecursiveT: EqualT: RenderTreeT: S
...
       case j @ CStage.Join(_, _) =>
         reifyJoin[G](j, parent, lens, isNested)
+
+      case CStage.Project(_) =>
+        absurd
     }
   }
 
+  // Order sibling cartouches when rendering to minimize
+  // structure and provide stability for tests
```
Comment (djspiewak) [inline qsu/src/main/scala/quasar/qsu/minimizers/MergeCartoix.scala]: 👍

File: qsu/src/main/scala/quasar/qsu/minimizers/CStage.scala (hunk 2 above)
Comment (wemrysi) [reply qsu/src/main/scala/quasar/qsu/minimizers/CStage.scala]: ...and partiality.

Comment (djspiewak) [reply qsu/src/main/scala/quasar/qsu/minimizers/CStage.scala]: Pssshaaaawww

File: qsu/src/main/scala/quasar/qsu/minimizers/CStage.scala (hunk 3 above)
Comment (wemrysi) [reply qsu/src/main/scala/quasar/qsu/minimizers/CStage.scala]: We no longer use indexed based naming as it was relying on two computations of the indices always agreeing and ended up being brittle, so I abandoned that in favor of explicit names for the identities.

**Comment by wemrysi** [comment]
Huh, I swear I ran the tests just before PR, apparently not 😞 I think failures are just due to cartoix reordering, will fix.

File: qsu/src/main/scala/quasar/qsu/minimizers/MergeCartoix.scala (hunk 6 above)
Comment (wemrysi) [reply qsu/src/main/scala/quasar/qsu/minimizers/MergeCartoix.scala]: No, this is referring to the case where you have a query like
```
select x{_}.y[_], x{_:}.y[_]
```
the `x{_}` and `x{_:}` will coalesce correctly, however in the identity access branch, the rest of the cartouche needs to be based on the shift identity, not the value. When rendering out the shift, we always pass a lens to the shift value to downstream cartoix.

**Comment by wemrysi** [comment]
@djspiewak ready for another look

**Pull request merged**